content/courses/native-onchain-development/program-derived-addresses.md

```diff
@@ -22,258 +22,353 @@ description: "Get a deeper understanding of PDAs."
+                    Ok(address) => return Some((address, bump_seed[0])),
+                    Err(PubkeyError::InvalidSeeds) => (),
+                    _ => break,
+                }
+            }
+            bump_seed[0] -= 1;
+        }
+        None
+    }
+
+    // ...
+}
+```
 
-The `try_find_program_address` function then introduces the `bump_seed`. The
-`bump_seed` is a `u8` variable with a value ranging between 0 to 255. Iterating
-over a descending range starting from 255, a `bump_seed` is appended to the
-optional input seeds which are then passed to the `create_program_address`
-function. If the output of `create_program_address` is not a valid PDA, then the
-`bump_seed` is decreased by 1 and the loop continues until a valid PDA is found.
+We can see that the `try_find_program_address` calls the
+`create_program_address` method.
 
 ```rust
 pub fn try_find_program_address(seeds: &[&[u8]], program_id: &Pubkey) -> Option<(Pubkey, u8)> {
 
-    let mut bump_seed = [std::u8::MAX];
+    // ...
     for _ in 0..std::u8::MAX {
         {
-            let mut seeds_with_bump = seeds.to_vec();
-            seeds_with_bump.push(&bump_seed);
+            // `create_program_address` is called here
             match Self::create_program_address(&seeds_with_bump, program_id) {
-                Ok(address) => return Some((address, bump_seed[0])),
-                Err(PubkeyError::InvalidSeeds) => (),
-                _ => break,
+               //...
             }
         }
-        bump_seed[0] -= 1;
+       //...
     }
-    None
 
 }
+
 ```
 
-The `create_program_address` function performs a set of hash operations over the
-seeds and `program_id`. These operations compute a key, then verify if the
-computed key lies on the Ed25519 elliptic curve or not. If a valid PDA is found
-(i.e. an address that is _off_ the curve), then the PDA is returned. Otherwise,
-an error is returned.
+### 3. create_program_address
+
+The `create_program_address` method performs a hashing operation over the seeds
+and `program_id`. These operations compute a key and verify whether it lies on
+the Ed25519 elliptic curve. If a valid PDA is found (i.e., an address that is
+_off_ the curve), then either the PDA or an error is returned.
+
+The source code for `create_program_address`:
 
 ```rust
 pub fn create_program_address(
     seeds: &[&[u8]],
     program_id: &Pubkey,
 ) -> Result<Pubkey, PubkeyError> {
-
-    let mut hasher = crate::hash::Hasher::default();
+    if seeds.len() > MAX_SEEDS {
+        return Err(PubkeyError::MaxSeedLengthExceeded);
+    }
     for seed in seeds.iter() {
-        hasher.hash(seed);
+        if seed.len() > MAX_SEED_LEN {
+            return Err(PubkeyError::MaxSeedLengthExceeded);
+        }
     }
-    hasher.hashv(&[program_id.as_ref(), PDA_MARKER]);
-    let hash = hasher.result();
 
-    if bytes_are_curve_point(hash) {
-        return Err(PubkeyError::InvalidSeeds);
-    }
+    #[cfg(not(target_os = "solana"))]
```

Comment:
I was using the source code from the current solana-program code. I thought that maybe the previous code was using an earlier version of Solana program when the content was written. The code dosen't run on "solana VM " instead it is only run out on other architectures hence the feature flag in current Solana program code. I have removed this and replaced it with comments `//..` to prevent any confusion